usr/include/pkcs11types.h

```diff
@@ -1151,6 +1159,10 @@ typedef CK_ULONG CK_RV;
 /* CKR_FUNCTION_REJECTED is new for v2.20 */
 #define CKR_FUNCTION_REJECTED                 0x00000200
 
+/* New to v3.0 */
+#define CKR_TOKEN_RESOURCE_EXCEEDED           0x00000201
+#define CKR_OPERATION_CANCEL_FAILED           0x00000202
```

Comment:
CKR_OPERATION_CANCEL_FAILED is not used in your implementation below. Does that mean, that you support the session cancel also for dual function operations?

The standard also describes, that this value should be returned, if an operation can not be canceled. How do you handle these cases?

Reply:
Right, the code also supports canceling dual function operations. For dual function operations simply both session states are active, and thus can be canceled without any special handling.

In opencryptoki, canceling a session can not fail. If a operation is not active that is to be canceled, this should not be treated as an error, according to the standard.

So yes, adding these return codes is somewhat superfluous, but I wanted to have them in pkcs11types.h anyway, just for completeness.

Reply:
I agree, the error code has to be added in anyways. If it is not used, because it could not happen in your implementation, it is also ok with me.